Add match_without_allocating match option

jacobtkeio opened the pull request:
This PR adds a MATCH_WITHOUT_ALLOCATING match option to sched-fluxion-resource, flux ion-resource, the C and C++ REAPI, and resource-query. A request for MATCH_WITHOUT_ALLOCATING will return resources that a jobspec would have been matched with under MATCH_ALLOCATE, but it will not actually allocate those resources.
A caveat is that MATCH_WITHOUT_ALLOCATING will return the soonest-available match, even into the future, while MATCH_ALLOCATE will return failure if no match is possible starting exactly at the requested match time.
 
 Matching without allocating is a first step towards basic multi-cluster scheduling.
 A more useful next step is to find and report the longest available matching resources in a cluster, for which the likelihood that another cluster has available matching resources within the same timeframe is highest. Then, users can broadcast requests for the longest available matching resources on many clusters (without actually allocating the requests) and then choose suitable times and resources on which to actually allocate their multi-cluster job.

jacobtkeio committed:
ion: Add match_wo_alloc cmd to ion-rsrc

Problem: The match_wo_alloc match option is not accessible from flux
ion-resource.

Add a flux ion-resource match without_allocating command.

jacobtkeio committed:
ion: Add match_wo_alloc cmd to ion-rsrc

Problem: The match_wo_alloc match option is not accessible from flux
ion-resource.

Add a `flux ion-resource match without_allocating` command.

jacobtkeio committed:
testsuite: Add t4017 for MWOA[_FUTURE]

Problem: No tests cover the functionality of match_without_allocating
and match_without_allocating_future.

Add t4017, based on t4001, to test match_without_allocating and
match_without_allocating_future.
Format t4001 with indents.

jacobtkeio committed:
testsuite: Add t4017 for MWOA[_FUTURE]

Problem: No tests cover the functionality of match-without-allocating
and match-without-allocating-future.

Add t4017 to test match-without-allocating and
match-without-allocating-future.

jacobtkeio committed:
dfu_impl: Check for negative jobspec duration

Problem: Checks for negative jobspec duration typically occur before
submission to the traverser, but some code paths do not perform this
check.

Check for negative jobspec durations in dfu_impl. This is cheap and
centralizes the check.

jacobtkeio committed:
dfu: Prevent negative jobspec duration

Problem: libjobspec does not check whether submitted jobspec durations
are negative.

Prevent negative jobspec durations in jobmeta_t::build in dfu_impl.
This is a cheap and central way to perform the check. Eventually, a
check in libjobspec should supersede this.

jacobtkeio committed:
modules: Make rsrc_opt_t inherit from rsrc_prop_t

Problem: resource_match_opts.* has two related classes,
resource_opt_t and resource_prop_t, which both share
setters and getters for every option in the resource module.
resource_opt_t has a resource_prop_t member variable
and calls m_resource_prop.get_...() in its own
resource_opt_t::get_...() functions, and likewise for
its setters and is_set-ers. This causes half of
resource_match_opts.* to be boilerplate.

Solution: Make resource_opt_t inherit from rsrc_prop_t,
and replace all resource_opt_t function calls with the
inherited resource_prop_t implementations. Remove the
duplicated functions in resource_match_opts.*.

milroy committed:
reapi: add basic allocation unit tests

Problem: REAPI does not have any allocation unit testing.

Add unit tests of initialization and allocation via Catch2.

jacobtkeio committed:
C++ REAPI: Add CLI support for MWOA

Problem: The MATCH_WITHOUT_ALLOCATING (MWOA) match option is not
supported by the C++ REAPI.

Bypass job registration and counting logic for MWOA requests and add
usage comments.

jacobtkeio committed:
C++ REAPI: Add CLI support for match_wo_alloc

Problem: The match_wo_alloc match option is not accessible from the
C++ REAPI.

Add a match-without-allocating match option to the C++ REAPI.

jacobtkeio committed:
C REAPI: Add CLI+mod support for match_wo_alloc

Problem: The match_wo_alloc match option is not accessible from the C
REAPI.

Add a match-without-allocating match option to the C REAPI.

jacobtkeio committed:
C REAPI: Add CLI support for MWOA

Problem: The MATCH_WITHOUT_ALLOCATING (MWOA) match option is not
supported by the C REAPI.

Assign jobid `-1` to MWOA requests and add usage comments.

jacobtkeio committed:
rq: Add support for match_wo_alloc[_future]

Problem: The match_wo_alloc[-future] match options are not accessible
from the resource-query utility.

Add match without_allocating[_future] options to resource-query.

jacobtkeio committed:
rq: Add support for match_wo_alloc[_future]

Problem: The match_wo_alloc[-future] match options are not accessible
from the resource-query utility.

Add match without_allocating[_future] options to resource-query.
Treat wo-alloc-future requests as non-allocating requests in
resource-query.

jacobtkeio committed:
trav: Implement MATCH_WITHOUT_ALLOCATING_FUTURE

Problem: Fluxion cannot search for matching resources in the future
without also allocating or reserving them. However, this future match
search is useful, especially for planning a reservation across several
clusters.

Handle the MATCH_WITHOUT_ALLOCATING_FUTURE match option in the
traverser. This returns a match, possibly starting in the future,
without allocating it in the planner.

jacobtkeio committed:
trav: Implement match_without_allocating[_future]

Problem: Fluxion cannot search for matching resources without also
allocating them. However, it is useful to query for matches and
perform allocation separately, especially for scheduling across
several clusters.

Handle the match_without_allocating and
match_without_allocating_future match options in the DFU traverser.
These return matches without allocating them in the planner.

jacobtkeio committed:
rsrc: Handle match-without-allocating[-future]

Problem: The match-without-allocating[-future] match options are
accessible from the resource module, but the resource module treats
them as normal allocation requests. Therefore, both
match-without-allocating[-future] requests themselves and the
resulting queried job state will return ALLOCATED, which is
inaccurate.

Return a MATCHED state on match-without-allocating[-future] requests
to the resource module.
Additionally, return a jobid of -1 to those requests.

jacobtkeio committed:
rsrc: Handle match_without_allocating[_future]

Problem: The match_without_allocating[_future] match options are
accessible from the resource module, but such matches are inaccurately
marked as `ALLOCATED` in the response and in the resulting job state.

Return a `MATCHED` state instead.
Additionally, return a jobid of -1 that denotes that such matches do not
belong to any particular job. This jobid only affects the match output
and perf tracking, since the match_without_allocating[_future] match
options do not get tracked by jobid anyway.

jacobtkeio committed:
t4014: Format heredoc with tabs

Problem: The heredoc in 'loading job-validator conf works' does not
ignore tabs and does not indent new lines, violating the semi-standard
format of sharness tests.

Make the heredoc ignore tabs and indent every line.

jacobtkeio committed:
match_op: Add MATCH_WITHOUT_ALLOCATING option

Problem: The traverser has no way to handle match_without_allocating
requests because there is no match_op_t value for them.

Create one.

jacobtkeio committed:
match_op: Add MATCH_WITHOUT_ALLOCATING option

Problem: To gather as much information as possible about prospective
allocations, users may need to query match information without actually
allocating those matches. The traverser has no way to handle these kinds
of requests currently, partially because there is no match_op_t value
for them.

Create a MATCH_WITHOUT_ALLOCATING match_op_t value.

jacobtkeio committed:
rq: Use string_to_match_op to simplify cmd_match

Problem: Both cmd_match and cmd_match_multi in command.cpp switch on the
input command to return a match_op_t. This functionality has since been
centralized in resource/policy/base/match_op.h, so it is much simpler to
call the string_to_match_op function instead of leaving in the switch
statements.

Replace the switch statements in cmd_match* with calls to string_to_match_op and
match_op_is_valid.

jacobtkeio committed:
match_op: Add string_to_match_op

Problem: In many places in resource/, we switch on a string to create an
appropriate match_op_t. This is prone to error because misspellings are
more likely to occur across so many duplicated strings, and any changes
to the match_op_t enums need to be mirrored correctly across many files.

Create a string_to_match_op function in match_op.h to centralize all
string -> match_op_t mapping.

jacobtkeio committed:
resource_match: Make run accept a match_op

Problem: run () has many strcmps for match commands instead of using
the string_to_match_op function. This is prone to error because these
strings need to be written identically in multiple places.

Pass a match_op_t into run () instead of a const char *. Convert the
strings to match_op_ts in run_match instead.

jacobtkeio committed:
reapi_test_c: Add MWOA[_FUTURE] test

Problem: No test exists for the MATCH_WITHOUT_ALLOCATING[_FUTURE]
match_op_ts in the C REAPI.

Add one.

jacobtkeio committed:
reapi_test_c: Add MWOA[_FUTURE] test

Problem: No test exists for the MATCH_WITHOUT_ALLOCATING[_FUTURE]
match_op_ts in the C REAPI.

Add one.

jacobtkeio committed:
C++ REAPI: Exit when rq->traverser_run () != 0

Problem: reapi_cli_t::match_allocate checks for run () != 0, but it only
exits with -1 if errno == ENOMEM. It should always exit if run ()
returns a non-zero result.

Return -1 when rq->traverser_run() returns a non-zero result regardless
of the value of errno.

jacobtkeio committed:
reapi_test_cxx: Add MWOA[_FUTURE] test

Problem: No test exists for the MATCH_WITHOUT_ALLOCATING[_FUTURE]
match_op_ts in the C++ REAPI.

Add one.

jacobtkeio commented:
I have added a "within" parameter for the resource module, REAPI C/C++, resource-query, and test suite. The "within" parameter forces the traverser to return only matches that start within the given duration (i.e. start at $now + $within or earlier).

It only affects MATCH_ALLOCATE_ORELSE_RESERVE and MATCH_WITHOUT_ALLOCATING right now, since those are the only two match options that can return resource sets that begin in the future. However, I added the parameter to every match option so that future match options don't need to worry about adding it manually. Internally, it is implemented as a "latest" parameter to dfu_match_t::run(), which specifies the latest start time allowed for returned matches.

After a while I started wondering if it would be simpler to add a "within" parameter as a jobspec attribute instead, since this would cut down on the amount of option-handling code spread across resource, REAPI, and rq. We may still want the option-handling code in any case, but the implementation may ultimately be cleaner based off a jobspec attribute, perhaps including on the flux-core side of things.

jacobtkeio committed:
match_op: Add MATCH_WITHOUT_ALLOCATING_FUTURE op

Problem: The MATCH_WITHOUT_ALLOCATING match option gives some
information about prospective allocations, but it can only match
resources at the time of the query; users may want to find prospective
allocations/reservations in the future. The traverser has no way to
handle these kinds of requests currently, partially because there is no
match_op_t value for them.

Create a MATCH_WITHOUT_ALLOCATING_FUTURE match_op_t value.

jacobtkeio committed:
match_op: Add MATCH_WITHOUT_ALLOCATING_FUTURE op

Problem: The MATCH_WITHOUT_ALLOCATING match_op should give as much
information as possible about prospective allocations, but it only
checks for prospective allocations starting at the time of the query.
This ignores all the prospective allocations that start in the future,
so if a user wanted to find the next available reservation time for a
jobspec without actually reserving it, they would need to manually
scan future start times using MATCH_WITHOUT_ALLOCATING until they
found a match.

Add a MATCH_WITHOUT_ALLOCATING_FUTURE match_op that performs this
search automatically and returns the earliest matching prospective
allocation.

jacobtkeio committed:
ion: Add match_wo_alloc_future cmd to ion-rsrc

Problem: The match_wo_alloc_future option is not accessible from flux
ion-resource.

Add a `flux ion-resource match without_allocating_future` command.

jacobtkeio committed:
ion: Add match_wo_alloc_future cmd to ion-rsrc

Problem: The match_wo_alloc_future option is not accessible from flux
ion-resource.

Add a flux ion-resource match without_allocating_future command.

jacobtkeio committed:
C++ REAPI: Handle MWOA_FUTURE requests

Problem: The C++ REAPI behaves differently for matches with
and without allocation, but MATCH_WITHOUT_ALLOCATING_FUTURE
is not handled.

Handle MWOA_FUTURE the same as MWOA.
Add MWOA_FUTURE to usage comments.

jacobtkeio committed:
C++ REAPI: Add CLI support for MWOA_FUTURE

Problem: The C++ REAPI behaves differently for matches with
and without allocation, but MATCH_WITHOUT_ALLOCATING_FUTURE
is not handled.

Bypass job registration and counting logic for MWOA_FUTURE requests and
add usage comments.

jacobtkeio committed:
C REAPI: Add CLI support for MWOA_FUTURE

Problem: The C REAPI currently returns a jobid from its job counter to
MWOA_FUTURE requests, but MWOA[_FUTURE] requests should not have a valid
jobid.

Assign jobid `-1` to MWOA_FUTURE requests and add usage comments.

jacobtkeio committed:
C REAPI: Handle MWOA_FUTURE requests

Problem: The C REAPI currently returns a jobid from its job counter to
MWOA_FUTURE requests, but MWOA[_FUTURE] requests should not have a valid
jobid.

Return jobid=-1 from MWOA_FUTURE requests to the C REAPI.
Also add a description of MWOA_FUTURE to match_allocate's usage comment.

jacobtkeio committed:
reapi_test_cxx: Test graph non-mutation under MWOA

Problem: None of the tests check if match_without_allocating[_future]
requests change the internal state of the resource graph, and we
want to make sure that they absolutely do not.

Add checks for graph equality based on the graphs' `idata` and
`schedule` information to the C++ REAPI test case for MWOA requests.

jacobtkeio committed:
reapi_test_cxx: Test graph non-mutation under MWOA

Problem: None of the tests check if match_without_allocating[_future]
requests change the internal state of the resource graph, and we
want to make sure that they absolutely do not.

Add checks for graph equality based on the graphs' `idata` and
`schedule` information to the C++ REAPI test case for MWOA requests.

jacobtkeio committed:
pool_infra_t: add equality check ignoring colors

Problem: pool_infra_t's equality method currently checks whether the
object's "colors" subsystem_key_vec is equal to the other object's.
However, comparing the raw uint64_t values does not capture the actual
colors being represented, and pool_infra_t objects do not currently
have a way to check what their uint64_t colors mean with respect to the
traverser's m_color object. Finally, all colors get reset to white at
the start of a traversal anyway, so pool_infra_ts with all else equal
besides their color values will still act the same in practice.

Add a second has_equal_behavior member function that checks equality
between pool_infra_t objects except for their colors.
Refactor operator== to use has_equal_behavior to reduce code
duplication.

jacobtkeio committed:
dfu: Stop implicit subplan init from [] w/ try_at

Problem: Previously, the traverser used the index operator to access
nodes' subplans, like `m_graph[u].subplans[s]`. Because the traversal to
prime the pruning filters during graph initialization does not touch
every node (it skips leaf nodes, which will never have pruning filters),
generalized logic to access node subplans using `[]` in subsequent
traversals can implicitly generate elements of the node's subplan list.
This is not an issue for correctness since the elements are initialized
to NULL, and the traverser always checks whether an element is null
before using it.
However, this causes the graph's state to change in cases where it
shouldn't. Specifically, this causes infra_data idempotence checks to
fail, due to a change in the length of the subplan list, after running a
MATCH_WITHOUT_ALLOCATING request, which does not actually add or remove
any subplans.

Replace all non-creational instances of the index operator `[s]` with
`.try_at(s)`, and check for boost::none instead of NULL in element
existence checks.

jacobtkeio committed:
reapi_cli: Make error block reachable w/ trycatch

Problem: A `job_info_t` initialization block in reapi_cli_impl.hpp was
checking if the return value of `std::make_shared` was `nullptr`. Since
make_shared throws `std::bad_alloc` instead of returning `nullptr`, this
block could never have run.

Instead, wrap the call to `std::make_shared` in a try/catch block for
`std::bad_alloc`.

jacobtkeio committed:
reapi_cli: Make error block reachable w/ trycatch

Problem: A `job_info_t` initialization block in reapi_cli_impl.hpp was
checking if the return value of `std::make_shared` was `nullptr`. Since
make_shared throws `std::bad_alloc` instead of returning `nullptr`, this
block could never have run.

Instead, wrap the call to `std::make_shared` in a try/catch block for
`std::bad_alloc`.

jacobtkeio committed:
reapi_test_cxx: Test match_op <-> str idempotence

Problem: No tests exist for the utility functions that convert between
match_op_ts and strings, and these functions are prone to error due to
misspellings in strings.
Similarly, no test exists for `match_op_valid`.

Add a test to reapi_test_cxx.cpp that checks idempotence running each
match_op_t through both `match_op_to_string` and `string_to_match_op`.
Also test that `match_op_valid` returns true for valid match_ops and
false for invalid match_ops.

jacobtkeio committed:
match_op: Add bounding enum values

Problem: Without START_ and END_ enum values to help write loops
through the match_op_t enum, tests for match_op_t must state all
match_op_t values explicitly. This is easy to forget when adding a new
match_op_t, and it causes quite a bit of code duplication.

Add START_ and END_ bounding values to the match_op_t enum.

jacobtkeio committed:
match_op: Query match_op<->string by a defined map

Problem: match_op_t's conversion functions are written with many
duplicated strings. This makes them hard to maintain, since each time
we add a new match option we need to write its equivalent string at
least twice.

Instead, keep a map between each match_op_t and its equivalent char *,
and rewrite the conversion functions to make use of this mapping.

This requires the addition of a match_op.cpp function and the linking of
qmanager, which uses these conversions, against the resource library.

jacobtkeio committed:
reapi tests: replace relative paths with envvar

Problem: The Catch2 reapi tests use relative paths to refer to files
under the t/ directory. This works, but it is hard to read and maintain
(especially through changes to the project's directory structure).

Use the SHARNESS_TEST_SRCDIR envvar to get the path to the t/ directory
instead. It is set in resource/reapi/test/CTestTestfile.cmake.

jacobtkeio committed:
t: Test for MWOA[_F] double booking on overrun job

Problem: The match_without_allocating option (MWOA) introduces a new
alloc_type_t, `AT_NO_ALLOC`, which is affected anew by a previous bug
with `AT_ALLOC` where jobs could get scheduled on resources currently
occupied by a job running past its walltime.  Including `AT_NO_ALLOC` in
the guard against this bug does the trick, but there is no test for this
behavior under MWOA[_FUTURE].

Add regression test to ensure MWOA[_F] requests do not match resources
at the same time as jobs running past their walltime.

Additionally, add a test to confirm that, even when a job is overrunning
`now`, future matches (orelse_reserve, MWOAF) involving the overrunning
resources succeed. In other words, test that Fluxion is optimistic with
respect to overrunning jobs when matching in the future, assuming both
that currently overrunning jobs will end soon and that future or active
jobs will not overrun.

jacobtkeio committed:
dfu: Prevent MWOA multiple booking on overrun job

Problem: When a job runs past its walltime, its resource counts still
get removed from the graph at the scheduled end time. When this happens,
the traverser considers the unavailable nodes as available. If fluxion
only considers these available resource counts, a job can get scheduled
while the previous is still cleaning up, causing a multiple booking
error. To prevent this, dfu_impl manually checks each node's
`.schedule.allocations` list to see if the node is still allocated to
the previous job and fails to match if it is.
This check is guarded by the alloc_type_t of the match because it should
only apply to `AT_ALLOC`. However, matching_without_allocating (MWOA)
should also perform this check to avoid reporting nodes available before
the previous job actually terminates.

Add `AT_NO_ALLOC` to this check.
Additionally, check whether or not `meta.at == meta.now`.
Otherwise, `match without_allocating_future` can fail when the node is
currently allocated but should successfully match in the future.

jacobtkeio committed:
reapi_cli_test_cxx: Add MWOA[_F] conversion cases

Problem: There is no test that checks whether the
`MATCH_WITHOUT_ALLOCATING` or `MATCH_WITHOUT_ALLOCATING_FUTURE`
match_op_ts get converted to the correct strings through the `match_op`
library.

Add cases for these new match_op_ts to the match_op_t conversion test in
`reapi_cli_test_cxx`.

jacobtkeio committed:
trav: Add AT_NO_ALLOC allocation type

Problem: The traverser has no way to emit matched resources without also
allocating them.

Create and implement a new alloc_type_t, `AT_NO_ALLOC`, that emits
matched resources without allocating them.

jacobtkeio committed:
trav: Implement MATCH_WITHOUT_ALLOCATING

Problem: Fluxion cannot search for matching resources without also
allocating them. However, it is useful to query for matches and
perform allocation separately, especially for scheduling across
several clusters.

Handle the MATCH_WITHOUT_ALLOCATING match option in the traverser.
This returns a match without allocating it in the planner.

milroy commented on resource/traversers/dfu.cpp:
```diff
@@ -386,6 +394,10 @@ int dfu_traverser_t::run (Jobspec::Jobspec &jobspec,
         < 0)
         return -1;
 
+    // If matching without allocation, set meta.AT to prevent allocation in update
```
I think you mean this:
```suggestion
    // If matching without allocation, set alloc_type to prevent allocation in update
```

milroy commented on resource/modules/resource.cpp:
```diff
@@ -383,7 +383,10 @@ static void match_request_cb (flux_t *h, flux_msg_handler_t *w, const flux_msg_t
-    if (is_existent_jobid (ctx, jobid)) {
+    if (std::string ("without_allocating") == cmd
+        || std::string ("without_allocating_future") == cmd) {
+        jobid = -1;
```
If I recall correctly, this is the sentinel value we agreed on. If so, please document it in a code comment and expand the commit message to indicate how it will be used.

milroy commented on t/t4017-match-without-allocating.t:
```diff
@@ -0,0 +1,125 @@
+test_expect_success 'loading resource module with policy=low works' '
+    load_resource \
+load-file=${grug} prune-filters=ALL:core \
+load-format=grug subsystems=containment policy=low
```
Nit: indentation
```suggestion
    load-file=${grug} prune-filters=ALL:core \
    load-format=grug subsystems=containment policy=low
```

jacobtkeio replied:
I've thrown in the same change for `t4001` as well!

milroy commented on t/t4017-match-without-allocating.t:
```diff
@@ -0,0 +1,125 @@
+    test $(grep ALLOCATED <rq.out | wc -l) -eq 4 &&
+    test $(grep "No matching resources found" <rq.out | wc -l) -eq 1 &&
+    grep "SCHEDULED AT" rq.out | sed -ne "6p" | grep 3600
+'
```
I think this test belongs in `t3001` in a dedicated `match_without_allocating` section.

milroy commented on resource/traversers/dfu.cpp:
```diff
@@ -101,10 +102,11 @@ int dfu_traverser_t::request_feasible (detail::jobmeta_t const &meta,
-            : graph_end - 1;
+    const int64_t target_time = op == match_op_t::MATCH_ALLOCATE
+                                        || op == match_op_t::MATCH_WITHOUT_ALLOCATING
+                                        || op == match_op_t::MATCH_WITHOUT_ALLOCATING_FUTURE
```
Caught with Claude's help:
```suggestion
```
`MATCH_WITHOUT_ALLOCATING_FUTURE` is semantically like `MATCH_ALLOCATE_ORELSE_RESERVE`, but the ternary makes it equivalent to `MATCH_ALLOCATE` here. So when a cluster is fully occupied now, the precheck probes `meta.at`, gets "unavailable" on every node, returns `EBUSY` — and the future-seeking loop that would have found the match at, say, `t=3600` is never reached.

Example reproducer:

```bash
$ ./resource/utilities/resource-query -L t/data/resource/grugs/tiny.graphml -S CA
> match allocate t/data/resource/jobspecs/match_format/t3040_only_node.yaml
> match allocate t/data/resource/jobspecs/match_format/t3040_only_node.yaml
# cluster now fully node-allocated
> match allocate_orelse_reserve <node jobspec with constraints: {...}>
  → RESERVED at 3600                    (correct)
> match without_allocating_future <same jobspec>
  → "No matching resources found" (EBUSY)   (should match at 3600)
```
Note the jobspec must use the `constraints:` key (plural — jobspec.cpp:337); a singular `constraint:` silently lands in optional and takes the fast path, masking the bug. That masking is also why `t4017`'s MWOAF test (unconstrained jobspec) passes.

jacobtkeio replied:
I failed to reproduce this because I couldn't get the planner to report unavailable at now when all nodes were allocated, probably because I missed something obvious. Let me know if you would really like a reproducer and I'll keep working it out. Otherwise this is a simple fix!

milroy replied:
You didn't miss anything obvious. Try this jobspec:

```yaml
# constrained_node.yaml — t3040_only_node.yaml plus a constraint
version: 9999
resources:
  - type: node
    exclusive: true
    count: 1
attributes:
  system:
    duration: 3600
    constraints:
      hostlist: [ "node[0-1]" ]
tasks:
  - command: [ "app" ]
    slot: default
    count:
      per_slot: 1
 ```
Then try this:
```bash
$ ./resource/utilities/resource-query -L t/data/resource/grugs/tiny.graphml -S CA
> match allocate t/data/resource/jobspecs/match_format/t3040_only_node.yaml
  → ALLOCATED at Now
> match allocate t/data/resource/jobspecs/match_format/t3040_only_node.yaml
  → ALLOCATED at Now                # both nodes of tiny.graphml now allocated
> match allocate_orelse_reserve constrained_node.yaml
  → RESERVED at 3600                # correct: future walk finds t=3600
> match without_allocating_future constrained_node.yaml
  → "No matching resources found"   # WRONG — should be MATCHED at 3600
```

jacobtkeio replied:
It worked like a charm! Thank you so much

milroy commented on resource/traversers/dfu_impl.cpp:
```diff
@@ -133,7 +133,9 @@ int dfu_impl_t::by_excl (const jobmeta_t &meta,
         // resources at the leaf level this check will not catch
         // multiple booking.
-        if (meta.alloc_type == jobmeta_t::alloc_type_t::AT_ALLOC
+        if (at == meta.now
```
The correct-but-too-expensive way to do this is:

```C++
  if (exclusive_in || resource.exclusive == Jobspec::tristate_t::TRUE) {
      // Reject vertices holding an overrun job: its planner span has
      // expired but the job has not been freed, so its true release
      // time is unknown. The planner alone would report such a vertex
      // available (issue #1043). Applies at every probe time: an overrun
      // job must not be offered now or at any future point.
      if (meta.alloc_type == jobmeta_t::alloc_type_t::AT_ALLOC
          || meta.alloc_type == jobmeta_t::alloc_type_t::AT_NO_ALLOC) {
          planner_t *plans = (*m_graph)[u].schedule.plans;
          for (auto const &[jobid, span] : (*m_graph)[u].schedule.allocations) {
              int64_t start = planner_span_start_time (plans, span);
              int64_t dur = planner_span_duration (plans, span);
              if (start < 0 || dur < 0 || start + dur <= meta.now) {
                  errno = EBUSY;
                  return -1;
              }
          }
      }
```
I don't think that even buys much given that MWOAF will be used to estimate future availability anyway. I recommend extending the current comment to state the intended behavior explicitly: at `now`, results are exact (overrun-safe); at future probe times, MWOAF is deliberately optimistic with respect to overrunning jobs. This raises another issue: I think MWOA/F should be documented in a README.

jacobtkeio replied:
I am having trouble understanding how this implementation differs in practice from the one in the PR. Would you mind explaining the issue this implementation solves?
I will definitely extend the comment to explain the optimism issue. Does allocate_orelse_reserve also behave in the same optimistic fashion? I think this case is equivalent to when a job right before an existing reservation runs over its wall time into the reservation time. Since we have no way to predict which jobs will overrun, aren't all future matches inherently optimistic?

milroy replied:
We discussed this in detail during our 1:1, but here's an explanation for posterity:

With node A overrunning its allocation plus other nodes with active spans that create a future point `t_e` (NB: depends on resources tracked by aggregate filter) the MWOAF walk will inspect `t_e`, and at `t_e > meta.now` the `at == meta.now` gate is skipped on every vertex including A. MWOAF can then report node A available at `t_e` while its overrunning job may still hold it. For internal reservations this optimism is self-correcting (qmanager re-schedules), but MWOAF results are transmitted externally (no re-scheduling feedback loop). The recommendation is just to document the behavior.

milroy commented on t/t1032-without-alloc-check.t:
```diff
@@ -0,0 +1,71 @@
+
+# Adapted from t1024
+
+test_description='
+Check that match without-allocating does not book resources when a previous
+allocation exceeds its walltime
+'
```
Claude ran this test with and without the previous commit, and the test passes regardless. Here's a summary:

The reason is that jobs are submitted as `flux submit -N1 -x hostname` with no -t limit → duration 0 → planner span extends to graph end → during the epilog the span is still active, so MWOA's EBUSY comes from the ordinary planner-busy path, not the new by_excl gate. The test never enters the overrun window this test claims to exercise.

Fix: mirror t1024 — `flux submit -N1 -x -t1s sleep 30`, wait for the exception `type=timeout` event (span now expired) then assert MWOA/MWOAF return 16 while the epilog runs. My probe confirmed this scenario distinguishes fixed from unfixed builds.

Also consider a mixed-cluster MWOAF test for ≥2 nodes, one overrunning, one with an active future span; assert MWOAF does offer the overrunning node at the other node's completion point.

jacobtkeio replied:
Fixed, and I confirmed it fails before the fix. Sorry about that! I'll add the mixed cluster test

milroy replied:
No problem, that's what reviews are for!

milroy reviewed (commented):
The PR is in pretty good shape already, and should be good to merge as soon as a few items are corrected and implemented.

jacobtkeio committed:
testsuite: Add rq test for MWOA[_FUTURE]

Problem: No test exists for `match without_allocating` or `match
without_allocating_future` requests to resource-query.

Add a basic one to `t3001-resource-basic.t`.

jacobtkeio committed:
doc: Add basic info a out MWOA[_F] to rq README

Problem: match_without_allocating[_future] is not documented in any
README.

Add basic descriptions of MWOA[_F] to resource-query's README.
Include information about MWOA[_F]'s jobid `-1` and MWOAF's optimism
with respect to overrunning jobs for matches in the future.

jacobtkeio committed:
t: Ensure MWOAF feasible if all allocated now

Problem: A bug in `request_feasible` that made without_allocate_future
return infeasible if all nodes were allocated at `now` has no regression
test.

Add one (thanks to @milroy!).
The test states that MWOAF should be able to match in the future even if
all nodes are allocated now.

milroy reviewed (approved):
Thanks for the patience and the diligent work on this feature, @jacobtkeio! This capability will be really useful for single- and multi-cluster workloads.

Also, credit to Claude for helping with the reproducer for MWOF feasible.

trws reviewed (approved):
Agreed, let's get this in. =)

jameshcorbett reviewed (approved):
I looked through briefly and am on board, I set MWP.
